src/V3HierBlock.cpp

```diff
@@ -0,0 +1,390 @@
+    }
+
+public:
+    Visitor(V3HierBlockPlan* planp, AstNetlist* netlist)
+        : m_planp(planp)
+        , m_modp(NULL) {
+
+        iterateChildren(netlist);
+    }
+    VL_DEBUG_FUNC;  // Declare debug()
+};
+
+//######################################################################
+
+V3HierBlockPlan::V3HierBlockPlan() {}
+
+bool V3HierBlockPlan::isHierBlock(const AstNodeModule* modp) const {
+    return m_blocks.find(modp) != m_blocks.end();
+}
+
+void V3HierBlockPlan::add(const AstNodeModule* modp, const std::vector<AstVar*>& gparams) {
+    iterator it = m_blocks.find(modp);
+    if (it == m_blocks.end()) {
+        V3HierBlock* hblockp = new V3HierBlock(modp, gparams);
+        UINFO(3, "Add " << modp->prettyNameQ() << " with " << gparams.size() << " parameters"
+                        << std::endl);
+        m_blocks.insert(std::make_pair(modp, hblockp));
+    }
+}
+
+void V3HierBlockPlan::registerUsage(const AstNodeModule* parentp, const AstNodeModule* childp) {
+    iterator parent = m_blocks.find(parentp);
+    UASSERT_OBJ(parent != m_blocks.end(), parentp, "must be added");
+    iterator child = m_blocks.find(childp);
+    if (child != m_blocks.end()) {
+        UINFO(3, "Found usage relation " << parentp->prettyNameQ() << " uses "
+                                         << childp->prettyNameQ() << std::endl);
+        parent->second->addChild(child->second);
+        child->second->addParent(parent->second);
+    }
+}
+
+void V3HierBlockPlan::createPlan(AstNetlist* nodep) {
+    // When processing a hierarchy block, no need to create a plan anymore.
+    if (v3Global.opt.hierMode()) return;
+
+    AstNodeModule* modp = nodep->topModulep();
+    if (modp->hierBlock()) {
+        modp->v3warn(HIERBLOCK, "Top module is marked hierachy block. Ignore the setting.");
+        modp->hierBlock(false);
+    }
+
+    vl_unique_ptr<V3HierBlockPlan> planp(new V3HierBlockPlan());
+    { Visitor visitor(planp.get(), nodep); }
+    V3Stats::addStat("HierBlock, Hierarchy block", v3Global.opt.hierBlocks().size());
+
+    // No hierarchy block is found, nothing to do.
+    if (planp->empty()) return;
+
+    v3Global.hierPlanp(planp.release());
+}
+
+V3HierBlockPlan::HierVector V3HierBlockPlan::hierBlocksSorted() const {
+    typedef std::map<const V3HierBlock*, std::set<const V3HierBlock*> > ChildrenMap;
+    ChildrenMap childrenOfHierBlock;
+
+    HierVector sorted;
+    for (const_iterator it = begin(); it != end(); ++it) {
+        if (!it->second->hasChild()) {  // No children, already leaf
+            sorted.push_back(it->second);
+        } else {
+            ChildrenMap::value_type::second_type& childrenSet
+                = childrenOfHierBlock[it->second];  // insert
+            const V3HierBlock::HierBlockSet& c = it->second->children();
+            childrenSet.insert(c.begin(), c.end());
+        }
+    }
+
+    // Use index instead of iterator because new elements will be added in this loop
```

Comment:
Consider if cleaner to add to new list instead?  Fine if leave as is too.

Also please make sure isn't a O(n^2) algorithm.

Reply:
Outer loop:
`i` increases monotonically from 0 to n-1 where `n` is the number of hierarchy blocks.

Inner loop:
The inner loop iterates `p` times and its cost is `log(p)` for associative container (map/set) access.
`p` is the number of parent hierarchy blocks of a hierarchy block. So `p log(p)`.

Using unordered container may decrease the complexity of the inner loop to `p`,  but I 'd like to keep as is until the complexity is visible as a performance cost. I recently found an [article](https://chromium.googlesource.com/chromium/src/base/+/master/containers/#Map-and-set-selection) from Chromium team.